### PR TITLE
fix: sort valid waypoints

### PR DESCRIPTION
**Pull request by allroundexperts**
### Details

This PR adds sorting logic to the validate waypoint function.

### Fixed Issues

$ https://github.com/Expensify/App/issues/26583
PROPOSAL: https://github.com/Expensify/App/issues/26583#issuecomment-1703949037

### Tests

1. Click on the `+` icon in the LHN.
2. Select "Request money" and then choose "Distance".

**Case 1**
1. Select the starting and finishing points
2. Click on "Add stop". Don't save an address for this new waypoint/stop.
3. Proceed by clicking the "Next" button.

Verify that the route is shown correctly.

**Case 2**
1. Select the starting and finishing points that are across continents.

Verify that an error occurs and the next button is disabled.

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

1. Click on the `+` icon in the LHN.
2. Select "Request money" and then choose "Distance".

**Case 1**
1. Select the starting and finishing points
2. Click on "Add stop". Don't save an address for this new waypoint/stop.
3. Proceed by clicking the "Next" button.

Verify that the route is shown correctly.

**Case 2**
1. Select the starting and finishing points that are across continents.

Verify that an error occurs and the next button is disabled.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/30054992/ee6ce280-17d5-4140-83e5-7f21dbf035be

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/30054992/0eec7e87-c45b-49d1-b453-99b1b634e1f5

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/30054992/4bc029ff-f483-4017-a3a4-51d7ef9e8d7b

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/30054992/a3cdc12a-a64b-4620-b66e-7f6736ea62e8

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/30054992/ffbfb2ab-7bb8-458c-8f0c-20cacab602dc

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/30054992/f740f190-2c0b-467a-a52c-0c3ccc7088c2

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
@allroundexperts @situchan could you resume the work on this please?

**Comment by situchan** [comment]
Waiting for @allroundexperts to complete checklist

**Comment by situchan** [comment]
Please also add error case in QA step

**Comment by situchan** [comment]
> disabled the next button when the user is offline

@hayata-suenaga please confirm if you also 👍

**Comment by trjExpensify** [comment]
Cool, so if you can't submit the form because of an API call that's required then a disabled confirmation button with the offline indicator underneath it is correct in the realm of pattern C. 

@neil-marcellini @hayata-suenaga I am curious though, I thought it was designed to be able to do the entire distance request flow offline using pattern B.

File: src/libs/TransactionUtils.js [hunk 1]
```diff
@@ -309,7 +309,8 @@ function waypointHasValidAddress(waypoint) {
  * @returns {Object} validated waypoints
  */
 function getValidWaypoints(waypoints, reArrangeIndexes = false) {
-    const waypointValues = _.values(waypoints);
+    const sortedIndexes = _.map(_.keys(waypoints), (key) => parseInt(key.replace('waypoint', ''), 10)).sort();
```
Comment (situchan) [inline src/libs/TransactionUtils.js:312]: There are multiple places where this logic is used. Can we make util function?
There are 3 usages of `const index = Number(key.replace('waypoint', ''));`

Comment (hayata-suenaga) [reply src/libs/TransactionUtils.js:312]: agree, and we eventually have to change the index logic for waypoints...

**Comment by hayata-suenaga** [comment]
> Probably, another follow up polish item we can do is to disable the next button when the getRoute api call is loading.

Is it hard to include this fix in your PR? 🙇

**Comment by hayata-suenaga** [comment]
> @hayata-suenaga What about this? Shall we create a follow up ticket?

I gonna create a follow up GH ticket for this. This is out of scope for this PR

We eventually want to handle this case -> user create waypoints on the Distance Request page while offline -> Click "Continue" (your PR temporarily prevents the user doing this) -> While on the confirmation page, the device comes back online -> route is fetched -> turned out that waypoints are invalid -> display error message on the confirmation page

soooo... if the user clicks the "Continue" button while the route is being loaded and that route turned out to be invalid, we might be able to handle that case on the confirmation page (when the above flow is supported)

**Comment by hayata-suenaga** [comment]
@situchan please continue with your checklist 🙇

**Comment by situchan** [comment]
Mostly looks good.

Let's fix this edge case:
Deep link to http://localhost:8082/request/new/confirmation/ directly after error shows.

I think the expected behavior is to redirect to /request/new/distance page as there's invalid waypoint.
Or leave this page and disable Request button?
cc: @hayata-suenaga 

https://github.com/Expensify/App/assets/108292595/83a8a3b1-be8f-4beb-8e73-0d64828a6490

**Review by neil-marcellini — COMMENTED** [review]
I think we should show an error message and prevent navigating to the next page when the finish is blank. Let's instruct the user to edit or remove the blank waypoint. Otherwise it's weird on the confirmation page that we only have start and stop waypoint icons, but no finish.

**Comment by hayata-suenaga** [comment]
> I think the scope is this PR is getting way too broad. It was meant to just sort the waypoints. Do we want to fix this in this PR?

Nop. We will handle an error when we submit the distance request

> When submitting a request, the last stop automatically becomes the ending point. I think its fine as it is.

So when there is an error, the user is prevented from submitting the distance request right?

**Comment by hayata-suenaga** [comment]
The deep link thing is causing so much headache. We need to do a refactor to not support deep linking for money request flow

**Comment by hayata-suenaga** [comment]
>> So when there is an error, the user is prevented from submitting the distance request right?
> 
> Yes. An error on the waypoint selection screen will disable the next button.

@allroundexperts 

could you kindly check what happens if there is an error on the confirmation screen? Can you still submit the request?

**Comment by situchan** [comment]
> Unexpected error requesting money, please try again later

https://github.com/Expensify/App/assets/108292595/b1154592-350f-4f55-b463-d4b58ba0be5d

**Comment by hayata-suenaga** [comment]
@situchan thank you!

This is something we might want to cleanup but out of scope for this PR

but I wanted to check the current behavior. Thank you so much!

**Comment by situchan** [comment]
I am not sure what's happening but in android I cannot get error (invalid waypoint) at all and finally app crashes after confirm request. No issue on all other platforms.


https://github.com/Expensify/App/assets/108292595/974bebf8-e4e6-44d2-ba8c-ebdb15200c31

**Comment by situchan** [comment]
> > I am not sure what's happening but in android I cannot get error (invalid waypoint) at all and finally app crashes after confirm request. No issue on all other platforms.
> > android.mov
> 
> Is your server set as staging?

I tested against both staging/production server

**Comment by situchan** [comment]
@allroundexperts can you test android with 3 locations I chose?

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/108292595/8dfe512f-5a8d-4111-b3c8-29a24a501901



</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/108292595/577d5178-35b7-4da3-a6d0-3a4fa7fe4b6f



</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/108292595/564a6c3a-3276-46b4-81ce-5a8301b2e199



</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/108292595/4ff8c48b-940b-442d-a470-d5d865116ccd



</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/108292595/cda2553d-2e7b-4bf2-ac71-ad90c0c6b4ee



</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/108292595/9979b851-e032-4a5d-97db-b30c12cf9bb5



</details>

**Review by situchan — APPROVED** [review]
:shipit:

**Comment by hayata-suenaga** [comment]
> I am not sure what's happening but in android I cannot get error (invalid waypoint) at all and finally app crashes after confirm request. No issue on all other platforms.

@situchan so this issue is not present?

**Review by hayata-suenaga — APPROVED** [review]

**Comment by hayata-suenaga** [comment]
Decided to hold this PR until the conference where we announce the feature is over 🙇 

Good work by the way 🎉

**Comment by neil-marcellini** [comment]
The conference and promotion is over so I'm taking this off hold and I'll give it another review now.

**Review by neil-marcellini — APPROVED** [review]

**Pull request merged**